proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -10611,86 +10628,87 @@ lemma refill_budget_check_valid_refills[wp]:
-   apply (intro hoare_vcg_conj_lift_pre_fix
-          ; (solves schedule_used_simple)?)
+   apply (intro hoare_vcg_conj_lift_pre_fix; (solves schedule_used_simple)?)
 
-          apply (find_goal \<open>match conclusion in "\<lbrace>_\<rbrace> head_insufficient_loop _ \<lbrace>\<lambda>_. _\<rbrace>" \<Rightarrow> \<open>-\<close>\<close>)
-          apply (rule_tac Q'="\<lambda>_ s. \<not> round_robin sc_ptr s \<and> sp_valid_refills_unbundled sc_ptr s"
-                       in hoare_strengthen_post[rotated])
-           apply (fastforce simp: valid_refills_def vs_all_heap_simps)
-          apply (intro hoare_vcg_conj_lift_pre_fix
-                 ; (solves head_insufficient_loop_simple)?)
-                apply (wpsimp wp: head_insufficient_loop_refills_sum)
-                apply (clarsimp simp: vs_all_heap_simps word_le_nat_alt)
-               apply (wpsimp wp: head_insufficient_loop_ordered_disjoint)
-               apply (clarsimp simp: vs_all_heap_simps word_le_nat_alt)
-              apply (wpsimp wp: head_insufficient_loop_no_overflow)
-              apply (clarsimp simp: vs_all_heap_simps word_le_nat_alt)
-             apply (wpsimp wp: head_insufficient_loop_refills_window)
-             apply (clarsimp simp: vs_all_heap_simps word_le_nat_alt)
-            apply (wpsimp wp: head_insufficient_loop_MIN_BUDGET_in_head)
-            apply (clarsimp simp: vs_all_heap_simps word_le_nat_alt)
-           apply (wpsimp wp: head_insufficient_loop_length_bounded)
-           apply (clarsimp simp: vs_all_heap_simps word_le_nat_alt)
-          apply (wpsimp wp: head_insufficient_loop_non_zero_refills)
-          apply (clarsimp simp: vs_all_heap_simps word_le_nat_alt)
-         apply (wpsimp wp: refill_budget_check_refills_sum_helper)
-        apply (wpsimp wp: refill_budget_check_ordered_disjoint_helper', fastforce+)
-       apply (wpsimp wp: refill_budget_check_no_overflow', fastforce+)
-      apply (wpsimp wp: refill_budget_check_window_helper, fastforce+)
-     apply (wpsimp wp: refill_budget_check_length_helper)
-    apply (wpsimp wp: refill_budget_check_non_zero_refills_helper)
-   apply (wpsimp wp: refill_budget_check_refills_unat_sum_helper, fastforce+)
-  apply (clarsimp simp: obj_at_def)
-  apply (wpsimp wp: set_refills_wp get_refills_wp
-              simp: vs_all_heap_simps update_sched_context_set_refills_rewrite update_refill_hd_rewrite)+
+           apply (find_goal \<open>match conclusion in "\<lbrace>_\<rbrace> head_insufficient_loop _ \<lbrace>\<lambda>_. _\<rbrace>" \<Rightarrow> \<open>-\<close>\<close>)
+           apply (rule_tac Q'="\<lambda>_ s. \<not> round_robin sc_ptr s \<and> sp_valid_refills_unbundled sc_ptr s"
+                        in hoare_strengthen_post[rotated])
+            apply (fastforce simp: valid_refills_def vs_all_heap_simps)
+           apply (intro hoare_vcg_conj_lift_pre_fix; (solves head_insufficient_loop_simple)?)
+                 apply (wpsimp wp: head_insufficient_loop_refills_sum)
```

Comment:
this looks like you can combine the rest of the proof into a `(wpsimp wp: ... | clarsimp simp: ...)+`, I don't think there's much of interest here for the proof length